implementation/building.tex

```diff
@@ -0,0 +1,165 @@
+By making continuous integration part of the normal build routine and review-process for new project code, we had an extra step of verification that the project would build on other systems then the one it was developed on.
+
+It is important to note that continuous integration only helped us to verify that the project could be \emph{built}, it could not help us to verify that the compiled code would actually \emph{work} for its target architecture.
+In order to verify that the code would work for on the Cortex-M3 we had to actually run in on one of the microcontrollers that we had a available for this project.
+An experimental process of testing and mocking the \texttt{emlib} \gls{api} bindings is described in greater detail in \autoref{ssub:testing}.
+
+\subsection{Contributing to Cargo}
+\label{ssub:contributing_to_cargo}
+
+As already mentioned, the ability to pass arbitrary flags down to the invocation of \texttt{rustc} was by design not supported by Cargo, but many people in the Rust community have wanted the ability to do so.
+The reasoning for not allowing arbitrary flags to be passed down is pretty obvious.
+A compilation can go awry very quickly if it is up to the package \emph{author} what flags should be passed to \texttt{rustc} instead of the package \emph{builder}.
+This would give the author the ability to set the restriction for a library, not the user.
+Different systems do not necessarily support all flags and possibilities, so if a package dependency says that it is to be built in a specific way it might not work for the system that is building it.
+\todo{Should we include references to the issues?}
+
+There have been an issue on Cargo's project page about implementing a new subcommand (called \texttt{cargo-rustc}) for the package manager.
+This subcommand would allowed for passing arbitrary flags further on to \texttt{rustc}, but with the restriction of only compiling a \emph{single} target at a time.
+This means that only \emph{either} the library, a binary, an example or a test (or a package dependency), may be compiled with the extra flags, and \emph{not} the entire package.
+These rules are restrictive enough to get libraries to not depend on a set of extra flags, but loose enough so that specialized projects, like our \texttt{emlib}, can depend on it for completing the build.
+Indeed, the functionality proposed with this subcommand was exactly the same as was solved with our implementation of \texttt{cargo-linkargs}.
+
+After gaining insight into Cargo's internals during the development of \texttt{cargo-linkargs}, it was interesting to see if we could get this same functionality into Cargo itself by implementing \texttt{cargo-rustc}.
+Not only would it benefit our project, it would also give many other Rust projects the ability to use Cargo for the entire compilation process.
```

Comment:
Emphasis the issues with depending on a plugin subcommand for cargo => not ergonomic.